src/bors/mergeable_queue.rs

```diff
@@ -0,0 +1,316 @@
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Item {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.expiration.cmp(&other.expiration)
+    }
+}
+
+struct SharedInner {
+    queue: Mutex<BinaryHeap<Item>>,
+    notify: Notify,
+    closed: AtomicBool,
+}
+
+#[derive(Clone)]
+pub struct MergeableQueueSender {
+    inner: Arc<SharedInner>,
+}
+
+#[derive(Clone)]
+pub struct MergeableQueueReceiver {
+    inner: Arc<SharedInner>,
+}
+
+pub fn create_mergeable_queue() -> (MergeableQueueSender, MergeableQueueReceiver) {
+    let shared = Arc::new(SharedInner {
+        queue: Mutex::new(BinaryHeap::new()),
+        notify: Notify::new(),
+        closed: AtomicBool::new(false),
+    });
+
+    (
+        MergeableQueueSender {
+            inner: shared.clone(),
+        },
+        MergeableQueueReceiver { inner: shared },
+    )
+}
+
+impl MergeableQueueSender {
+    pub fn close(&self) {
+        self.inner.closed.store(true, Ordering::Relaxed);
+        self.inner.notify.notify_waiters();
+    }
+
+    pub fn enqueue(&self, repo: GithubRepoName, pr_number: PullRequestNumber) {
+        let expiration = Some(Instant::now() + BASE_DELAY);
+
+        self.insert_item(
+            MergeableQueueItem {
+                pull_request: QueuedPullRequest { pr_number, repo },
+                attempt: 1,
+                immediate: false,
+            },
+            expiration,
+        );
+    }
+
+    #[allow(dead_code)]
+    pub fn enqueue_now(&self, pr_number: PullRequestNumber, repo: GithubRepoName) {
+        self.insert_item(
+            MergeableQueueItem {
+                pull_request: QueuedPullRequest { pr_number, repo },
+                attempt: 1,
+                immediate: true,
+            },
+            None,
+        );
+    }
+
+    pub fn enqueue_retry(&self, queue_item: MergeableQueueItem) {
+        let next_attempt = queue_item.attempt + 1;
+        let delay = calculate_exponential_backoff(BASE_DELAY, BACKOFF_MULTIPLIER, next_attempt);
+        let expiration = Some(Instant::now() + delay);
+
+        self.insert_item(
+            MergeableQueueItem {
+                pull_request: queue_item.pull_request,
+                attempt: next_attempt,
+                immediate: queue_item.immediate,
+            },
+            expiration,
+        );
+    }
+
+    fn insert_item(&self, item: MergeableQueueItem, expiration: Option<Instant>) {
+        let mut queue = self.inner.queue.lock().unwrap();
+
+        // Notify when:
+        // 1. The current item expires sooner than the head of the queue
+        let has_earlier_expiration =
+            queue
+                .peek()
+                .is_some_and(|head| match (expiration, head.expiration) {
+                    (Some(new_exp), Reverse(Some(head_exp))) => new_exp < head_exp,
+                    _ => false,
+                });
+        // 2. The queue was empty before insertion (reader is waiting)
```

Comment:
```suggestion
        // 2. The queue was empty before insertion (reader might be waiting)
```
There might not be any reader.